src/Lecture4.hs

```diff
@@ -226,7 +288,7 @@ implement the next task.
 -}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
+combineRows = sconcat . fmap rowToStats
```

Comment:
The second fix needed for implementing the optimal lazy solution is to use `foldl'` instead of `sconcat`.

The implementation of `sconcat` is the following:

```haskell
sconcat :: Semigroup a => NonEmpty a -> a
sconcat (a :| as) = go a as
  where
    go b (c:cs) = b <> go c cs
    go b []     = b
```

As you can notice, this implementation is not tail-recursive and not strict. So it leads to space leaks.

Instead, you can implement a helper function like:

```haskell
combineStats :: NonEmpty Stats -> Stats
```

It's pretty straightforward to implement it using pattern matching on the only `NonEmpty` constructor and `foldl'`

Reply:
it seems I still need more improvements... memory usage grows until 8 GB after a couple of minutes of running 😱

Reply:
Well, I pushed the suggestion about using a custom function to combine Maybes and that made the trick. I have now a constant memory usage of about 16 Mb. 🎉